android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/internal/SyncPerformanceTests.kt

```diff
@@ -0,0 +1,536 @@
+    ): List<Document>? {
+
+        val user = client.auth.user ?: return null
+        val array: List<Byte> = (0 until sizeOfDocsInBytes).map { 0.toByte() }
+        val docs: List<Document> = (0 until numberOfDocs).map {
+            Document(mapOf(
+                "_id" to ObjectId(),
+                "owner_id" to user.id,
+                "bin" to Binary(array.toByteArray())
+            ))
+        }
+        return docs
+    }
+
+    @Before
+    override fun setup() {
+        super.setup()
+
+        if (!Stitch.hasAppClient(stitchOutputAppName)) {
+            outputClient = Stitch.initializeAppClient(stitchOutputAppName)
+            outputClient.auth.loginWithCredential(AnonymousCredential())
+
+            outputColl = outputClient
+                .getServiceClient(RemoteMongoClient.factory, "mongodb-atlas")
+                .getDatabase(stitchOutputDbName)
+                .getCollection(stitchOutputCollName)
+        }
+    }
+
+    private fun setupIter() {
+        Assume.assumeTrue("no MongoDB URI in properties; skipping test", getMongoDbUri().isNotEmpty())
+        setup()
+
+        val app = createApp()
+
+        addProvider(app.second, ProviderConfigs.Anon)
+        mdbService = addService(
+                app.second,
+                "mongodb",
+                "mongodb1",
+                ServiceConfigs.Mongo(getMongoDbUri())).second
+
+        dbName = ObjectId().toHexString()
+        collName = ObjectId().toHexString()
+        namespace = MongoNamespace(dbName, collName)
+
+        val rule = RuleCreator.MongoDb(
+                database = dbName,
+                collection = collName,
+                roles = listOf(RuleCreator.MongoDb.Role(
+                        read = true, write = true
+                )),
+                schema = RuleCreator.MongoDb.Schema())
+        mdbRule = addRule(mdbService, rule)
+
+        this.client = getAppClient(app.first)
+        userId = Tasks.await(client.auth.loginWithCredential(AnonymousCredential())).id
+        mongoClient = client.getServiceClient(RemoteMongoClient.factory, "mongodb1")
+        coll = mongoClient.getDatabase(dbName).getCollection(collName)
+        (mongoClient as RemoteMongoClientImpl).dataSynchronizer.stop()
+        (mongoClient as RemoteMongoClientImpl).dataSynchronizer.disableSyncThread()
+        BaseStitchAndroidIntTest.testNetworkMonitor.connectedState = true
+    }
+
+    @After
+    override fun teardown() {
+        super.teardown()
+    }
+
+    private fun teardownIter() {
+        val syncedIds = Tasks.await(coll.sync().syncedIds)
+        Tasks.await(coll.sync().desyncMany(*syncedIds.toTypedArray()))
+        Tasks.await(coll.deleteMany(Document()))
+
+        if (::mongoClient.isInitialized) {
+            (mongoClient as RemoteMongoClientImpl).dataSynchronizer.close()
+            AndroidEmbeddedMongoClientFactory.getInstance().close()
+        }
+
+        teardown()
+    }
+
+    private fun runPerformanceTestWithParams(
```

Comment:
Can possibly extrapolate some of the logic to private methods so that it is more legible.

Reply:
Will work on this